### PR TITLE
Adding Splunk output-types for fluentbit

### PR DESCRIPTION
**Pull request by momoXD007**
### What this PR does / why we need it:
This PR adds the ability to fluent-operator to configure Splunk as an output for fluentbit. Fluentbit already supports Splunk as output  but so far fluent-operator simply can not configure a fluentbit-pod to actually make use of that. This PR intends to change that. 
In other words it enables users of the fluent-operator to use more of the features fluentbit already has.

I am not an experienced go-developer and have simply used similar PRs (e.g.: PRs that add the opensearch-output types) and fluentbits' documentation for Splunk [(here)](https://docs.fluentbit.io/manual/pipeline/outputs/splunk) as base for my work.
Please review and test carefully.

### Which issue(s) this PR fixes:

Fixes #

### Does this PR introduced a user-facing change?

```release-note
None
```

### Additional documentation, usage docs, etc.:

```docs
- [How to configure Splunk in fluentbit] https://docs.fluentbit.io/manual/pipeline/outputs/splunk
```

### CODE REVIEW

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// OpenSearch is the opensearch output plugin, allows to ingest your records into an OpenSearch database.
+type Splunk struct {
+	// IP address or hostname of the target OpenSearch instance, default `127.0.0.1`
+	Host string `json:"host,omitempty"`
+	// TCP port of the target Splunk instance, default `8088`
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=65535
+	Port *int32 `json:"port,omitempty"`
+	// Specify the Authentication Token for the HTTP Event Collector interface.
+	SplunkToken string `json:"splunk_token,omitempty"`
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:21]: Maybe `splunkToken` is better

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	// Specify X-Splunk-Request-Channel Header for the HTTP Event Collector interface.
+	Channel string `json:"channel,omitempty"`
+
+	// Optional username credential for access
+	HTTPUser *plugins.Secret `json:"httpUser,omitempty"`
+	// Password for user defined in HTTP_User
+	HTTPPasswd *plugins.Secret `json:"httpPassword,omitempty"`
+	// If the HTTP server response code is 400 (bad request) and this flag is enabled, it will print the full HTTP request
+	// and response to the stdout interface. This feature is available for debugging purposes.
+	HTTPDebugBadRequest bool `json:"httpDebugBadRequest,omitempty"`
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:36]: `*bool` maybe better,like this https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/plugins/output/elasticsearch_types.go#L61

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	// When enabled, the record keys and values are set in the top level of the map instead of under the event key. Refer to
+	// the Sending Raw Events section from the docs for more details to make this option work properly.
+	SplunkSendRaw bool `json:"splunkSendRaw,omitempty"`
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:39]: `*bool`

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	if o.HTTPPasswd != nil {
+		pwd, err := sl.LoadSecret(*o.HTTPPasswd)
+		if err != nil {
+			return nil, err
+		}
+		kvs.Insert("HTTP_Passwd", pwd)
+	}
+	if o.Compress != "" {
+		kvs.Insert("Index", o.Compress)
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:100]: Index?

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	if o.Channel != "" {
+		kvs.Insert("Type", o.Channel)
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:103]: `Type` is wrong.

**Review by wenchajun — CHANGES_REQUESTED** [review]

**Comment by benjaminhuo** [comment]
@wanjunlei would you help to review this?
@wenchajun What else should @momoXD007 do to fix the helm e2e error? https://github.com/fluent/fluent-operator/actions/runs/3203685595/jobs/5236604040

**Comment by wenchajun** [comment]
These two commands should be executed after the changes are made.
```
make manifests
make generate
```

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	if o.Host != "" {
+		kvs.Insert("Host", o.Host)
+	}
+	if o.Port != nil {
+		kvs.Insert("Port", fmt.Sprint(*o.Port))
+	}
+	if o.SplunkToken != "" {
+		kvs.Insert("Splunk_Token", o.SplunkToken)
+	}
+	if o.BufferSize != "" {
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:82]: should be `http_buffer_size`

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+// Name implement Section() method
+func (_ *Splunk) Name() string {
+	return "splunk"
+}
+
+// Params implement Section() method
+func (o *Splunk) Params(sl plugins.SecretLoader) (*params.KVs, error) {
+	kvs := params.NewKVs()
+	if o.Host != "" {
+		kvs.Insert("Host", o.Host)
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:74]: Uppercase is not recommended for code consistency.

kvs.Insert("Host", o.Host) => kvs.Insert("host", o.Host)

Others also need to be modified.

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	if o.Compress != "" {
+		kvs.Insert("Index", o.Compress)
+	}
+	if o.Channel != "" {
+		kvs.Insert("Type", o.Channel)
+	}
+	if o.HTTPDebugBadRequest {
+		kvs.Insert("HTTP_Debug_Bad_Request", "On")
+	} else {
+		kvs.Insert("HTTP_Debug_Bad_Request", "Off")
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:108]: It is redundant, the default is `Off`.

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	if o.SplunkSendRaw {
+		kvs.Insert("Splunk_Send_Raw", "On")
+	} else {
+		kvs.Insert("Splunk_Send_Raw", "Off")
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:113]: same as above

**Review by wanjunlei — CHANGES_REQUESTED** [review]

**Comment by momoXD007** [comment]
I hopefully fixed all findings n ow :) Thanks for looking into this so quickly

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+		kvs.Insert("host", o.Host)
+	}
+	if o.Port != nil {
+		kvs.Insert("port", fmt.Sprint(*o.Port))
+	}
+	if o.SplunkToken != "" {
+		kvs.Insert("splunk_token", o.SplunkToken)
+	}
+	if o.BufferSize != "" {
+		kvs.Insert("buffer_size", o.BufferSize)
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:83]: buffer_size => http_buffer_size

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+	Host string `json:"host,omitempty"`
+	// TCP port of the target Splunk instance, default `8088`
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=65535
+	Port *int32 `json:"port,omitempty"`
+	// Specify the Authentication Token for the HTTP Event Collector interface.
+	SplunkToken string `json:"splunkToken,omitempty"`
+	//Buffer size used to receive Splunk HTTP responses: Default `2M`
+	// +kubebuilder:validation:Pattern:="^\\d+(k|K|KB|kb|m|M|MB|mb|g|G|GB|gb)?$"
+	BufferSize string `json:"bufferSize,omitempty"`
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:24]: bufferSize -> http_buffer_size

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	// Set payload compression mechanism. The only available option is gzip.
+	Compress string `json:"compress,omitempty"`
+	// Specify X-Splunk-Request-Channel Header for the HTTP Event Collector interface.
+	Channel string `json:"channel,omitempty"`
+
```
Comment (wenchajun) [inline apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:29]: There is no need for a blank line here.

File: apis/fluentbit/v1alpha2/plugins/output/splunk_types.go (hunk 11 above)
Comment (momoXD007) [reply apis/fluentbit/v1alpha2/plugins/output/splunk_types.go:24]: ah, ok that's what you meant, will fix that :)

**Review by wenchajun — APPROVED** [review]

**Review by wanjunlei — APPROVED** [review]

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**